sync15-adapter/src/state.rs

```diff
@@ -46,3 +73,568 @@ impl GlobalState {
+                self.state.collections = collections;
+                Ok(InitialWithLiveTokenAndInfo)
+            }
+
+            // Compare local and remote `meta/global` timestamps to determine
+            // if our locally cached `meta/global` is up-to-date.
+            InitialWithLiveTokenAndInfo => {
+                let action = {
+                    let local = self.state.global.as_ref().map(|global| &global.modified);
+                    let remote = self.state.collections.get("meta");
+                    FetchAction::from_modified(local, remote)
+                };
+                Ok(match action {
+                    // Hooray, we don't need to fetch `meta/global`. Skip to
+                    // the next state.
+                    FetchAction::Skip => HasMetaGlobal,
+                    // Our `meta/global` is out of date, or isn't cached
+                    // locally, so we need to fetch it from the server.
+                    FetchAction::Fetch => {
+                        let previous_global = mem::replace(&mut self.state.global, None);
+                        NeedsFreshMetaGlobal { previous_global }
+                    }
+                    // We have a `meta/global` record in our cache, but not on
+                    // the server. This likely means we're the first client to
+                    // sync after a node reassignment. Invalidate our cached
+                    // `meta/global` and `crypto/keys`, and try to fetch
+                    // `meta/global` from the server anyway. If another client
+                    // wins the race, we'll fetch its `meta/global`; if not,
+                    // we'll fail and upload our own.
+                    FetchAction::InvalidateThenUpload => {
+                        let previous_global = mem::replace(&mut self.state.global, None);
+                        self.state.keys = None;
+                        NeedsFreshMetaGlobal { previous_global }
+                    }
+                })
+            }
+
+            // Fetch `meta/global` from the server.
+            NeedsFreshMetaGlobal { previous_global } => match self.client.fetch_meta_global() {
+                Ok(new_global) => Ok(ResolveMetaGlobal {
+                    previous_global,
+                    new_global,
+                }),
+                Err(err) => match err.kind() {
+                    ErrorKind::NoMetaGlobal { .. } => Ok(FreshStartRequired { previous_global }),
+                    _ => Err(err),
+                },
+            },
+
+            // Reconcile the server's `meta/global` with our locally cached
+            // `meta/global`, if any.
+            ResolveMetaGlobal {
+                previous_global,
+                new_global,
+            } => {
+                // If the server has a newer storage version, we can't
+                // sync until our client is updated.
+                if &new_global.payload.storage_version > &STORAGE_VERSION {
```

Comment:
Why all the references?  Can't you just `new_global.payload.storage_version > STORAGE_VERSION`?